### DISCUSSION TITLE
QueryNode CPU usage is not balanced

### DISCUSSION THREAD

**Discussion by dutysloth** [milvus-io/milvus#32324] [Q&A and General discussion]
Hi
I Have a Milvus Cluster (v2.3.13) with 10 QueryNodes (20CPU 48GB), and I noticed that the CPU usage is really unbalanced between them.
While applying a relatively heavy search load, 2 QNs reach ~20CPU, while the other 8 QNs range from ~3-6CPU.
Memory-wise the QNs are quite balanced (~18-23GB).
I also ran a check to see the segments distribution and it is balanced.

Any ideas?
Thanks

**Comment by yhmo** [comment]
What is the shards_num value of the collection? Use `print(collection.describe()) `to print the collection's info.

A possible reason is that some of the segments are not indexed, the index tasks are not finished. Then milvus does brute-force search on these unindexed segments.

**Reply by dutysloth** [reply]
The shards_num is 2
I checked what you suggested using `utility.index_building_progress()` and it seems that everything is indexed (pending_index_rows=0).

**Reply by yhmo** [reply]
Show me the index parameters, and search parameters, vector dimension.

**Reply by dutysloth** [reply]
We have an 'IVF_FLAT' index (IP, nlist=2048) on a 512 dimension vector column.
There is also another index on a int16 column, if it matters.
In search we use nprobe=512. I'm aware it's really high but we require very high recall.

**Reply by xiaofan-luan** [reply]
there is no reason you use nlist = 2048 with such large nprobe.

If you need high recall you should go for HNSW

**Reply by xiaofan-luan** [reply]
for nlist =2048, typical nprobe is 64 or 128

**Reply by dutysloth** [reply]
I understand that the nprobe is maybe not ideal and that it causes slower performance, but is it really the reason for the QNs being unbalanced?
I mean the balance should keep regardless of how I search (which index, what params etc..), no?

**Comment by xiaofan-luan** [comment]
@dutysloth 
I thought that is because of delegator takes too much cpu on search.

This is something we want to improve, and we can definitely improve.

Please send me email at xiaofan.luan@zilliz.com and we can do further discussion?

**Comment by xiaofan-luan** [comment]
since you have two shards, each shard delegator takes not only growing search but also forward all the request, which may takes more cpus

**Comment by xiaofan-luan** [comment]
This can be fixed in 3.0 with the new architecture. But so far I can help on figuring out a way to improve

**Comment by xiaofan-luan** [comment]
The reason is because querynode play two roles, serving segment and shard delegator.
Since you only have two shards, only two of the nodes works as shard delegator. 
Shard delegator is responsible for:
1. forward request to querynodes
2. search on growing segments.
We need to figure out why this takes too much cpus. A perf or pprof might be really helpful
